src/OrchardCore.Modules/OrchardCore.Users/Services/ExternalLoginFormEvents.cs

```diff
@@ -0,0 +1,64 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Options;
+using OrchardCore.Mvc.Core.Utilities;
+using OrchardCore.Users.Controllers;
+using OrchardCore.Users.Events;
+using OrchardCore.Users.Models;
+
+namespace OrchardCore.Users.Services;
+
+public sealed class ExternalLoginFormEvents : LoginFormEventBase
+{
+    private readonly ExternalLoginOptions _externalLoginOptions;
+    private readonly SignInManager<IUser> _signInManager;
+    private readonly LinkGenerator _linkGenerator;
+    private readonly IHttpContextAccessor _httpContextAccessor;
+
+    public ExternalLoginFormEvents(
+        IOptions<ExternalLoginOptions> externalLoginOptions,
+        SignInManager<IUser> signInManager,
+        LinkGenerator linkGenerator,
+        IHttpContextAccessor httpContextAccessor)
+    {
+        _externalLoginOptions = externalLoginOptions.Value;
+        _signInManager = signInManager;
+        _linkGenerator = linkGenerator;
+        _httpContextAccessor = httpContextAccessor;
+    }
+
+    public override async Task<IActionResult> LoggingInAsync()
```

Comment:
Note: if you don't want your event to return a result and/or prefer applying the challenge immediately (rather than letting MVC do it later), you can execute the challenge immediately using `IAuthenticationService.ChallengeAsync()` instead of using `ChallengeResult`.

Reply:
@kevinchalet thanks for this tip. In the Login action, how would we check if the challenge was applied and what would we return from the action that implements IActionResult?